READY TO REVIEW: Relationship generator plugin

leftclickben opened the pull request:
Not sure how this took <s>32</s> <s>48</s> <s>70</s> 74 commits!  (Although to be fair, some of it was WIP commits to move from one machine to another - WAM, Gaia, home - and some was merging changes from upstream).

leftclickben committed:
WIP commit

leftclickben committed:
WIP commit

leftclickben committed:
WIP commit

leftclickben committed:
WIP commit

leftclickben committed:
WIP commit

leftclickben committed:
Merge branch 'RelationshipGeneratorPlugin' of github.com:leftclickben/providence into RelationshipGeneratorPlugin

leftclickben committed:
Fix merge issues.

leftclickben committed:
Fix merge issues.

leftclickben committed:
Fix merge issues.

leftclickben committed:
Fix merge issues.

leftclickben committed:
WIP relationship generator plugin

leftclickben committed:
WIP relationship generator plugin

leftclickben committed:
WIP relationship generator plugin

leftclickben committed:
WIP relationship generator plugin

leftclickben committed:
Working version of relationshipGeneratorPlugin.

Config contains dummy triggers/rules, some cleanup still required.

leftclickben committed:
Working version of relationshipGeneratorPlugin.

Config contains dummy triggers/rules, some cleanup still required.

leftclickben committed:
Working version of relationshipGeneratorPlugin.

Config contains dummy triggers/rules, some cleanup still required.

leftclickben committed:
Working version of relationshipGeneratorPlugin.

Config contains dummy triggers/rules, some cleanup still required.

Commit:
WIP Relationship generator plugin

Another WIP checkin to change machines.

Commit:
WIP Relationship generator plugin

Another WIP checkin to change machines.

Commit:
WIP Relationship generator plugin

Another WIP checkin to change machines.

Commit:
WIP Relationship generator plugin

Another WIP checkin to change machines.

leftclickben committed:
Documentation and name changes.

+ Add code documentation.
+ Rename "triggers" to "rules".
+ Rename "assign_on_[insert|update]" to "process_on_[insert|update]".
+ Add "add_matched" config item for better configurability.

leftclickben committed:
Documentation and name changes.

+ Add code documentation.
+ Rename "triggers" to "rules".
+ Rename "assign_on_[insert|update]" to "process_on_[insert|update]".
+ Add "add_matched" config item for better configurability.

leftclickben committed:
Documentation and name changes.

+ Add code documentation.
+ Rename "triggers" to "rules".
+ Rename "assign_on_[insert|update]" to "process_on_[insert|update]".
+ Add "add_matched" config item for better configurability.

leftclickben committed:
Documentation and name changes.

+ Add code documentation.
+ Rename "triggers" to "rules".
+ Rename "assign_on_[insert|update]" to "process_on_[insert|update]".
+ Add "add_matched" config item for better configurability.

leftclickben committed:
Fix incorrect variable naming convention.

leftclickben committed:
Fix incorrect variable naming convention.

leftclickben committed:
Merge branch 'RelationshipGeneratorPlugin' of github.com:leftclickben/providence into RelationshipGeneratorPlugin

Conflicts:
	app/plugins/relationshipGenerator/relationshipGeneratorPlugin.php

leftclickben committed:
Merge branch 'RelationshipGeneratorPlugin' of github.com:leftclickben/providence into RelationshipGeneratorPlugin

leftclickben committed:
Merge branch 'master' of github.com:wamuseum/providence into RelationshipGeneratorPlugin

leftclickben committed:
Make plugin remove relationships correctly.

+ Fix detection of existing relationships.
+ Add code-level documentation.
+ Add `notify` config item.
+ Configuration of rules is still only temporary.

leftclickben committed:
WIP - handle multiple values

+ Saving WIP due to encountering issue with caProcessTemplateForIDs() function.
+ Converted some more magic strings to constants (sub-keys under "rules").

leftclickben committed:
Rewrite of match code for greater configurability.

+ Configurable combination operators (and, or) for combining values across
  many fields, and for combining multiple values within a single field.
+ Configurable match methods (regex, exact, caseInsensitive) globally, per
  rule and per trigger.
+ Easily extensible as operator and match methods are called dynamically.
+ Regex patterns specified in configuration no longer have delimiters.

leftclickben committed:
Improve comments.

leftclickben committed:
Update rules to test exact match operator.

leftclickben committed:
Make method static.

leftclickben committed:
Configurable notifications.

+ Allow default notification templates and overrides per-rule for notifications
  for both adding and removing relationships
+ Code tidy-up

leftclickben committed:
Documentation

leftclickben committed:
Configuration status check.

+ Implemented the checkStatus() method (and some internal methods that it calls)
+ Some minor code refactors

leftclickben committed:
Add missing configuration requirement.

leftclickben committed:
Add checks to prevent misconfiguration causing PHP errors.

Also add some missing configuration requirements.

leftclickben committed:
Add unit tests for relationship generator plugin - WIP

+ Add some configuration-check testing unit tests
+ Add configuration files for tests
+ Add plugins test suite to phpunit.xml

leftclickben committed:
Use better constants.

leftclickben committed:
Formatting

leftclickben committed:
Configuration check unit tests.

+ Check for valid and invalid configuration files.
+ Check for valid and invalid rules at the top-level, rule level and trigger level.
+ Rename test class to correctly reflect what it is testing.

leftclickben committed:
Fix test class name in comment

leftclickben committed:
Skeleton for RelationshipGeneratorPluginIntegrationTest

Method names and comments only so far.

leftclickben committed:
Lazy load NotificationManager

This prevents errors in unit/integration tests.

leftclickben committed:
Configuration for plugin integration test.

leftclickben committed:
Error preventing type check.

leftclickben committed:
Relationship generator plugin integration tests.

+ Added working setUp() and tearDown() methods
+ Added first two tests

leftclickben committed:
BUGFIX - avoid error when there are no representations of an object.

+ Simple if() wrapper to ensure the representations exist before iterating, this
  avoids "invalid foreach() parameter" errors

leftclickben committed:
Remove boilerplate rules

+ These were only temporary so they are being removed
+ The default ruleset is empty
+ The test rules will be in the integration test's config file

leftclickben committed:
Rename variable for correct naming conventions.

leftclickben committed:
WIP integration test

+ Complete setUp() / tearDown() cycle including *most* referenced records
+ Use a configuration template and process it to the final configuration file
+ Ignore the processed configuration file for integration test, it is generated

leftclickben committed:
WIP integration test

+ Remove processed config file from vcs, not sure how that got in there.

leftclickben committed:
Don't use hardcoded locale, and some code formatting.

+ Get the locale setting from __CA_DEFAULT_LOCALE__

leftclickben committed:
Fix issue with runs of multiple tests.

+ Share fixtures between tests and use static setUpBeforeClass() / tearDownAfterClass()
  methods instead of non-static setUp() / tearDown()
+ This commit only changes non-static to static and $this to self
+ Oh and added some phpdoc

leftclickben committed:
Correct naming convention for statics.

leftclickben committed:
Simplify, maaaan, simplifyyyyyy...

leftclickben committed:
Generated metadata elements

+ Create metadata elements in setup and remove in teardown
+ Use dynamic element codes

leftclickben committed:
Change name of collections and add test

leftclickben committed:
More flexible match type and operator specification

+ Allow "case insensitive" as well as "caseInsensitive"

leftclickben committed:
Many more unit tests

+ Add unit tests for various circumstances - adding and removing
  relationships in various combinations

leftclickben committed:
Remove unused config

leftclickben committed:
Generate relationship type

+ Remove all dependencies on configurable data

leftclickben committed:
Remove TODO

+ The test was previously leaving remnant data in the
  ca_metadata_type_restrictions table, but it doesn't seem to
  be doing that any more, so removing the TODO note about it

leftclickben committed:
Documentation and method rename.

leftclickben committed:
Remove commented-out boilerplate config.

kehh commented on app/plugins/relationshipGenerator/conf/relationshipGenerator.conf:
```diff
@@ -0,0 +1,51 @@
+
+# Rules describing how relationships are generated
+rules = {
+}
```
Were you going to add a commented out example rule here?

leftclickben replied:
I decided the unit tests were enough, I'll add a comment pointing in that direction

leftclickben committed:
Add comment about example rules.

Add comment in default plugin configuration file about finding valid example rules
in the configuration file for the integration test.

leftclickben committed:
Use correct variable naming convention.

+ Change camelCase to underscore (s/camel/_/)
+ Add some missing type prefixes

leftclickben committed:
Improve documentation.

+ Add comments to integration test configuration file template explaining the use of %%placeholders%%.
+ Ignore comment lines when processing configuration file during integration test.

leftclickben committed:
Rename variables for correct naming conventions.

s/camel/_/
